src/envs/cover.py

```diff
@@ -345,3 +345,372 @@ def __init__(self) -> None:
+        # env. The action space now is (dx, dy, dgrip). The
+        # last dimension controls the gripper "magnet" or "vacuum".
+        # Note that the bounds are relatively low, which necessitates
+        # multi-step options.
+        return Box(-0.1, 0.1, (3,))
+
+    def simulate(self, state: State, action: Action) -> State:
+        # Since the action space is lower level, we need to write
+        # a lower level simulate function.
+        assert self.action_space.contains(action.arr)
+        dx, dy, dgrip = action.arr
+        next_state = state.copy()
+        x = state.get(self._robot, "x")
+        y = state.get(self._robot, "y")
+        grip = state.get(self._robot, "grip")
+        # Update robot state based on action.
+        x += dx
+        y += dy
+        grip += dgrip
+        next_state.set(self._robot, "x", x)
+        next_state.set(self._robot, "y", y)
+        next_state.set(self._robot, "grip", grip)
+        # If y is below a threshold, then we need to be in hand region.
+        if y < self.hand_region_height:
+            hand_regions = self._get_hand_regions(state)
+            # If we're not in any hand region, no-op.
+            if not any(hand_lb <= x <= hand_rb
+                       for hand_lb, hand_rb in hand_regions):
+                return next_state
+        # Identify which block we're holding and which block we're above.
+        held_block = None
+        above_block = None
+        for block in self._blocks:
+            if state.get(block, "grasp") != -1:
+                assert held_block is None
+                held_block = block
+            block_x_lb = state.get(block, "x") - state.get(block, "width")/2
+            block_x_ub = state.get(block, "x") + state.get(block, "width")/2
+            block_y_lb = state.get(block, "y") - self.grasp_height_tol
+            block_y_ub = state.get(block, "y") + self.grasp_height_tol
+            if state.get(block, "grasp") == -1 and \
+               block_x_lb <= x <= block_x_ub and \
+               block_y_lb <= y <= block_y_ub:
+                assert above_block is None
+                above_block = block
+        # If we're not holding anything and we're above a block, grasp it.
+        # The grasped block's pose stays the same.
+        # Note: unlike parent env, we also need to check the grip.
+        if held_block is None and above_block is not None and \
+           grip > self.grasp_thresh:
+            grasp = x - state.get(above_block, "x")
+            next_state.set(above_block, "grasp", grasp)
+        # If we are holding something, place it.
+        # Disallow placing on another block or in free space.
+        # Note: unlike parent env, we also need to check the grip.
+        # Also unlike parent env, we disallow placing except when
+        # below the hand_region_height.
+        if held_block is not None and above_block is None and \
+            grip < self.grasp_thresh and y < self.hand_region_height:
```

Comment:
Don't you also have to make sure the block you are holding is not below the y-axis?

Reply:
(resolved in conversation)